src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs

```diff
@@ -401,6 +416,196 @@ static void PromoteCommonCommitCharactersOntoList(LSP.VSCompletionList completio
+            LSP.CompletionParams request,
+            Document document,
+            int position,
+            CompletionTrigger completionTrigger,
+            OptionSet completionOptions,
+            CompletionService completionService,
+            CompletionListCache completionListCache,
+            CancellationToken cancellationToken)
+        {
+            var completionList = await completionService.GetCompletionsAsync(document, position, completionTrigger, options: completionOptions, cancellationToken: cancellationToken).ConfigureAwait(false);
+            cancellationToken.ThrowIfCancellationRequested();
+            if (completionList == null || completionList.Items.IsEmpty)
+            {
+                return null;
+            }
+
+            // Cache the completion list so we can avoid recomputation in the resolve handler
+            var resultId = completionListCache.UpdateCache(request.TextDocument, completionList);
+
+            return (completionList, resultId);
+        }
+
+        private static (CompletionList CompletionList, bool IsIncomplete) FilterCompletionList(CompletionList completionList, TextSpan completionListSpan, CompletionTrigger completionTrigger, SourceText sourceText, Document document)
+        {
+            var filterText = sourceText.GetSubText(completionListSpan).ToString();
+
+            // Use pattern matching to determine which items are most relevant out of the calculated items.
+            using var _ = ArrayBuilder<MatchResult<CompletionItem?>>.GetInstance(out var matchResultsBuilder);
+            var index = 0;
+            var completionHelper = CompletionHelper.GetHelper(document);
+            foreach (var item in completionList.Items)
+            {
+                if (CompletionHelper.TryCreateMatchResult<CompletionItem?>(
+                    completionHelper,
+                    item,
+                    editorCompletionItem: null,
+                    filterText,
+                    completionTrigger.Kind,
+                    GetFilterReason(completionTrigger),
+                    ImmutableArray<string>.Empty,
+                    includeMatchSpans: false,
+                    index,
+                    out var matchResult))
+                {
+                    matchResultsBuilder.Add(matchResult);
+                    index++;
+                }
+            }
+
+            // Next, we sort the list based on the pattern matching result.
+            matchResultsBuilder.Sort(MatchResult<CompletionItem?>.SortingComparer);
+
+            // Finally, truncate the list to 1000 items plus any preselected items that occur after the first 1000.
+            var filteredList = matchResultsBuilder
+                .Take(MaxCompletionListSize)
+                .Concat(matchResultsBuilder.Skip(MaxCompletionListSize).Where(match => match.RoslynCompletionItem.Rules.SelectionBehavior == CompletionItemSelectionBehavior.HardSelection))
```

Comment:
Can you have more than one hard selection?

Reply:
potentially yes, this is the same as current implementation.

Reply:
Interesting, out of curiosity what does it mean from the end-users perspective to have more then one?

Reply:
The client just picks the first one.  Your question made me look into it though - I think we need to actually be looking for this rule - https://sourceroslyn.io/#Microsoft.CodeAnalysis.Features/Completion/MatchPriority.cs,32 but only when the item is also hard selected.  Since LSP doesn't support us specifying a soft-selected preselect, imho its better to just not preselect in that case.

Updated